### ISSUE TITLE
[no-invalid-this] allow `this` when it's explicitly declared in function arguments

### ISSUE THREAD

**Issue by bajtos** [typescript-eslint/typescript-eslint#604]
**Repro**

I created a new repository reproducing the issue, see https://github.com/bajtos/eslint-sandbox (commit https://github.com/bajtos/eslint-sandbox/commit/859fd818b15915f17ad993589ab52d8edb7eefe6).

```JSON5
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    // IT IS CRUCIAL TO SET sourceType,
    // OTHERWISE THE ISSUE CANNOT BE REPRODUCED
    "sourceType": "module",
  },
  "rules": {
    "no-invalid-this": "error"
  }
}
```

```TS
describe('foo', () => {
  it('does something', function(this: Mocha.Context) {
    this.timeout(100);
    // done
  });
});
```

**Expected Result**

No problems are reported.

**Actual Result**

```
$ eslint test.ts

/Users/bajtos/src/eslint-sandbox/test.ts
  3:5  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

I am not sure where to report this error, sorry if I picked a wrong repo or a wrong issue template. 

I am guessing this problem is caused by interaction between `no-invalid-this` implementation built inside `eslint` that's not aware of TypeScript's `this` argument notation, and `@typescript-eslint/parser` that is failing to communicate the fact that a free-form function is expecting `this` argument to be set by the caller.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/parser`        | `1.10.2` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `12.1.0` |
| `npm`                              | `6.9.0` |

**Comment by astoilkov** [comment]
An additional scenario for which `typescript-eslint` rule will be beneficial in favor of ESLint default `no-invalid-this` is when you access `this` during the initialization of class property where ESLint incorrectly flags it as error.

```typescript
private readonly _resizeScheduler: IAnimationFrameScheduler = createAnimationFrameScheduler(() =>
  /**
   * ESLint incorrectly finds it as error.
   */
  // eslint-disable-next-line no-invalid-this
  this._resizeLines(),
)
```

**Comment by petemcwilliams** [comment]
also in arrow functions
```
onClick: () => {
                    this._presentIframeDetailModal();
 }
```

**Comment by anikethsaha** [comment]
I think its not because of the function argument type notation for this.
I tested the base rule and its working fine 

![image](https://user-images.githubusercontent.com/26347874/77856547-876a6d00-7215-11ea-853e-a8f5c9cf624b.png)

> note that I have commented out the `this.timeout`

![image](https://user-images.githubusercontent.com/26347874/77856587-c7315480-7215-11ea-8215-24c040d5919b.png)

> error in `this.timeout

Same as in [eslint demo](https://eslint.org/demo#eyJ0ZXh0IjoiXCJ1c2Ugc3RyaWN0XCI7XG4gICAgICBkZXNjcmliZSgnZm9vJywgKCkgPT4ge1xuICAgICAgICBpdCgnZG9lcyBzb21ldGhpbmcnLCBmdW5jdGlvbigpIHtcbiAgICAgICAgIHRoaXMudGltZW91dCgxMDApO1xuICAgICAgICAgIC8vIGRvbmVcbiAgICAgICAgfSk7XG4gICAgICB9KTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWludmFsaWQtdGhpcyI6Mn0sImVudiI6e319fQ==)

And I think this is an intended behavior as it disallows `this` except for 

> -  outside of classes or class-like objects
> - The function is on an object literal.
> - The function is assigned to a property.
> - The function is a method/getter/setter of ES2015 Classes. (excepts static methods)
> - The `call/apply/bind` method of the function is called directly.
> - The function is a callback of array methods (such as `.forEach()`) if `thisArg` is given.
> - The function has `@this` tag in its `JSDoc` comment.

According to the [docs](https://eslint.org/docs/rules/no-invalid-this)

**Comment by bradzacher** [comment]
it should be valid in this case though.
Typescript allows you to annotate the type of the `this` for a function:
```ts
  interface SomeType { prop: string }
  function foo(this: SomeType) {
//             ^^^^^^^^^^^^^^ 
    this.prop;
  }
```

This is _not_ an actual arg for the function, it's a special, type-only argument.
This is saying to typescript "when called, this function should have a `this` context that is of the specified type".

---

In the example code, it _looks_ like the function should have no "this" context, which means the rule is flagging the usage of `this` as invalid.
```ts
it('does something', function() {
  this.timeout(100);
  // done
});
```

However, unknown to both typescript and eslint - when mocha calls the function, it calls it `.call`/`.apply`, and it binds the `this` to the current mocha context - which ofc means that the code is actually valid.

So the example code annotates the `this` type to inform typescript of the correct type to use:
```ts
it('does something', function(this: Mocha.Context) {
  this.timeout(100);
  // done
});
```

However, the eslint has no knowledge of how to deal with TS's special `this` parameter, so it just assumes the `this` is still invalid.

---

An extension rule should do the following things:
- when it encounters a `this` reference in a function body, it should interrogate the function's arguments to see if there is an annotated `this` parameter.
- when it encounters a `this` reference in a class property, it should ignore it.

**Comment by anikethsaha** [comment]
thanks for the explanation

> - when it encounters a this reference in a function body, it should interrogate the function's arguments to see if there is an annotated this parameter.

Should it validate the type ? like to check the type content and check whether the correct method or variable is being used or not.

something like this

```ts
interface SomeType {
  var1: string;
}
function foo(this: SomeType) {
  this.prop; // error
  this.var1; // ok
}
```

**Comment by bradzacher** [comment]
> like to check the type content and check whether the correct method or variable is being used or not.

No, typescript does this itself - we don't want to duplicate errors that typescript already checks.

**Comment by anikethsaha** [comment]
ok cool, I will try to fix it soon 👍

**Comment by masi** [comment]
The rule triggers also for event handlers:
```
element.addEventListener('mouseover', function (): void {
   initHover(this);
});

```
How can I use the rule without getting errors?
